### ISSUE TITLE
Proposal: remove some ipyparallel-specific functionality

### ISSUE THREAD

**Issue by minrk** [ipython/ipykernel#17]
Mainly the `apply` message-related bits. These aren't part of the wider Jupyter message protocol, and are wholly Python specific. These can be moved to ipyparallel, and attached to the Kernel in a subclass.

**Comment by takluyver** [comment]
No objections from me.

**Comment by minrk** [comment]
Missed the boat for 4.0 on this one, but I still think it's the right move in the long run.

**Comment by dsblank** [comment]
FYI, we use the ipyparallel bits in wrapper kernels (via metakernel) to run other languages in parallel. This is quite powerful, and hope that we can keep that functionality.

**Comment by minrk** [comment]
@dsblank how do you use those kernels in parallel? Are you using ipyparallel at all? I mainly mean to remove the native-serialization stuff (data-pub, apply), not anything else, which shouldn't affect any kernels other than IPython.

**Comment by minrk** [comment]
@dsblank I've implemented most of what I intended in ipython/ipyparallel#61. I don't understand enough about how your use of parallel in metakernel works, so I don't know how to tell what would be affected. Can you provide some examples that I can test with?

**Comment by dsblank** [comment]
Sorry for the delay. Here is a short example:

1. Install a metakernel-based kernel: `pip3 install metakernel_bash --user`
2. Start a cluster
3. Start the kernel
4. Start remote kernels: `%parallel metakernel_bash MetaKernelBash`
5. Execute a command: `%px uname`

Hopefully that will still work. Thanks!

**Comment by minrk** [comment]
@dsblank thanks, that still works after the above PR. Since it looks like you are sitting on top of the existing IPython engines and clients, moving the implementation shouldn't have any effect on your use case.

**Comment by dsblank** [comment]
Thanks for checking @minrk ! It is quite nice to have Parallel Java, Parallel Scheme, Parallel Cobol ...

**Comment by Carreau** [comment]
bump to 6.0 ?
